include/lsst/afw/table/io/FitsSchemaInputMapper.h

```diff
@@ -0,0 +1,160 @@
+    ) const = 0;
+
+    virtual ~FitsColumnReader() {}
+
+};
+
+/**
+ *  A structure that describes a field as a collection of related strings read from the FITS header.
+ */
+struct FitsSchemaItem {
+    int column;          // column number (0-indexed); -1 for Flag fields
+    int bit;             // flag bit number (0-indexed); -1 for non-Flag fields
+    std::string ttype;   // name of the field (from TTYPE keys)
+    std::string tform;   // FITS column format code (from TFORM keys)
+    std::string tccls;   // which field class to use (from our own TCCLS keys)
+    std::string tunit;   // field units (from TUNIT keys)
+    std::string doc;     // field docs (from comments on TTYPE keys)
+
+    explicit FitsSchemaItem(int column_, int bit_) : column(column_), bit(bit_) {}
+};
+
+
+/**
+ *  A class that describes a mapping from a FITS binary table and an afw::table Schema.
+ *
+ *  A FitsSchemaInputMapper is created every time a FITS binary table is read into an afw::table
+ *  catalog, allowing limited customization of the mapping between on-disk FITS table columns
+ *  an in-memory fields by subclasses of BaseTable.
+ *
+ *  The object is constructed from a daf::base::PropertyList that represents the FITS header,
+ *  which is used to populate a custom container of FitsSchemaItems.  These can then be retrieved
+ *  by name or column number via the find() methods, allowing the user to create custom readers
+ *  for columns or groups of columns via addColumnReader().  They can also be removed from the
+ *  "regular" fields via the erase() method.  Those regular fields are filled in by the finalize()
+ *  method, which automatically generates mappings for any FitsSchemaItems that have not been
+ *  removed by calls to erase().  Once finalize() has been called, readRecord() may be called
+ *  repeatedly to read FITS rows into record objects according to the mapping that has been
+ *  defined.
+ */
+class FitsSchemaInputMapper {
+public:
+
+    typedef FitsSchemaItem Item;
+
+    /// Construct a mapper from a PropertyList of FITS header values, stripping recognized keys if desired.
+    FitsSchemaInputMapper(daf::base::PropertyList & metadata, bool stripMetadata);
+
+    /**
+     *  Set the Archive to an externally-provided one, overriding any that may have been read.
+     */
+    void setArchive(PTR(InputArchive) archive);
+
+    /**
+     *  Set the Archive by reading from the HDU specified by the AR_HDU header entry.
+     *
+     *  Returns true on success, false if there is no AR_HDU entry.
+     */
+    bool readArchive(afw::fits::Fits & fits);
+
+    /// Return true if the mapper has an InputArchive.
+    bool hasArchive() const;
+
+    /**
+     *  Return a mutable reference to the Schema object that is being built.
+     *
+     *  This allows new fields to be added by the user (usually those required by custom FitsColumnReaders).
+     *  It should not be called after finalize() has been called.
+     */
+    Schema & editSchema();
```

Comment:
I don't think this is actually used anywhere at present. Is it necessary?

Would it be worth storing a flag to indicate whether the mapper has been finalized and throwing if it is set when ``editSchema()`` is called?